File: CWE122/C/heapBufferOverflow.c
/* To run native: 
    gcc heapBufferOverflow.c

   To run Web Assembly: 
    emcc -g heapBufferOverflow.c -o heapBufferOverflow.html
        1. In the js file enter your arguments in the "arguments_" array
        2. Launch local host server for the html file
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
# define BUFSIZE 8

void printHeapData(char *ptr1, char *ptr2) {
    printf("Heap Data 1: %s, Memory Address: %p\n", ptr1, ptr1);
    printf("Heap Data 2: %s, Memory Address: %p\n", ptr2, ptr2);
}

int main(int argc, char *argv[]) {
    
    printf("argv[1]: %s\n", argv[1]);
    printf("argv[2]: %s\n", argv[2]);

    char *ptr1, *ptr2;
    ptr1 = malloc(BUFSIZE);
    ptr2 = malloc(BUFSIZE);

    printf("\nExecute:\nchar *ptr1, *ptr2;\nptr1 = malloc(%i);\nptr2 = malloc(%i);\n\n", BUFSIZE, BUFSIZE);

    printHeapData(ptr1, ptr2);

    strcpy(ptr1, argv[1]);

    printf("\nExecute:\nstrcpy(ptr1, argv[1]);\n\n");

    printHeapData(ptr1, ptr2);

    strcpy(ptr2, argv[2]);

    printf("\nExecute:\nstrcpy(ptr2, argv[2]);\n\n");
    
    printHeapData(ptr1, ptr2);

    return 0;

}

File: CWE122/Go/heapBufferOverflow.go
/* To run native:
    go run heapBufferOverflow.go

   To run Web Assembly:
    GOOS=js GOARCH=wasm go build -o heapBufferOverflow.wasm heapBufferOverflow.go
        1. In the js file enter your arguments in the "this.argv" array on line 97
        2. Launch local host server for the html file
*/

package main

import (
	"fmt"
	"os"
)

const BUFSIZE = 8

func printHeapData(ptr1, ptr2 []byte) {
	fmt.Printf("Heap Data 1: %s, Memory Address: %p\n", ptr1, &ptr1[0])
	fmt.Printf("Heap Data 2: %s, Memory Address: %p\n", ptr2, &ptr2[0])
}

func main() {
	if len(os.Args) < 3 {
		fmt.Println("Please provide two command line arguments.")
		return
	}

	fmt.Printf("argv[1]: %s\n", os.Args[1])
	fmt.Printf("argv[2]: %s\n", os.Args[2])

	ptr1 := make([]byte, BUFSIZE)
	ptr2 := make([]byte, BUFSIZE)

	fmt.Printf("\nExecute:\nvar ptr1, ptr2 []byte\nptr1 = make([]byte, %d)\nptr2 = make([]byte, %d)\n\n", BUFSIZE, BUFSIZE)

	printHeapData(ptr1, ptr2)

	copy(ptr1, os.Args[1])

	fmt.Printf("\nExecute:\ncopy(ptr1, os.Args[1])\n")

	printHeapData(ptr1, ptr2)

	copy(ptr2, os.Args[2])

	fmt.Printf("\nExecute:\ncopy(ptr2, os.Args[2])\n")

	printHeapData(ptr1, ptr2)
}
